### DISCUSSION TITLE
[RFC] Bring an additional NestJS style API to Blitz

### DISCUSSION THREAD

**Discussion by paulm17** [blitz-js/blitz#4088] [RFCs]
Recently I have been speaking with @siddhsuresh about bringing an API that is similar to NestJS to BlitzJS.  He has asked me to raise an RFC so that the community could bring about any ideas or comments they have surrounding this topic.  Also it's inspired by https://github.com/blitz-js/blitz/discussions/1841.

Now I do wish to make clear the difference between the two threads.

- I am 100% advocating for an opinionated framework.
- It should not be for routes as this would increase complexity considerably and honestly there be dragons.  Because now that BlitzJS is a framework.  I think the easiest approach would be to implement an api endpoint.  Which could possibly spin off a future mechanism for NextJS middleware? 🤔 
- It would allow for the sharing of types between the NestJS API BE and NextJS BE.  Huge win.
- The same prisma schema used between both.  Those who currently use a monorepo would welcome this I'm sure.
- It would use the same existing node process.  Not a custom server.
- This add-on would then pave the way for:-

1. Better file uploads.  Including Tus resumable.
2. Websockets.
3. Cron job.
4. Queues.
5. Other security aspects, JWT creation, etc.
6. And more importantly, introduce DRY coding aspects and allow for complex business logic. 

Now don't get me wrong.  What NextJS supplies for the current API endpoint is ok.  It's just that what you can do with NestJS is that it's typically used for those use cases where heavy business logic is being implemented.  Every developer that I have spoken to on discord who has outgrown NextJS API has reached for NestJS and been able overcome obstacles due to NextJS.

If anyone has used the microservices version of nestJS.  This is pretty much what I am advocating for.  I think it should cover the majority of cases.

Finally, I came across this which should showcase how powerful having the same types across an FE and api BE could be: 

> https://ts-rest.com/

Also it could also point to ideas on how to start?  Maybe implement a simple version of express and go from there?  🤔 

Now over to you.  What do you think?  If you are using NestJS along with your BlitzJS project.  I'd definitely welcome your thoughts too.  Thanks!

**Comment by flybayer** [comment]
Overall I think this is a great idea! It just needs a champion with the time and vision to make it a reality

**Reply by jhonnymichel** [reply]
File uploads and Cron Jobs are two items that particularly interest me as I actually have had to implement separate solutions for those in Blitz apps. I can be part of the authoring team

**Comment by datner** [comment]
This sounds amazing, I'm exactly at the situation you're describing. Though we don't have to look at nestjs for direction. It's an amazing framework, seriously. But when it comes to it, it's built using the mental model for angular and C# developers, and it won't be in our favor if folks start comparing blitz and nestjs. 

Lately I've been getting into effect-ts really hard, it's beyond amazing. I'm using it currently with blitz and it's a dream. There are other new interesting consequences to a push like you're outlining. I'm not invested enough to champion this, but I would like to help and be kept in the loop for this

**Reply by Kay2dan** [reply]
Would love to hear more about this...

**Reply by datner** [reply]
@Kay2dan [the site (wip)](https://effect.website)

**Comment by Inalegwu** [comment]
I've been using NestJS for a while now and this would be an intresting thing to see but Nest follows the Angular model and despite Blitz pivoting towards a more framework agnoistic approach I feel these changes would be better suited as packages for the Blitz ecosystem or features for BlitzRPC instead of a new layer.
